MaxLines property on Label

jfversluis opened the pull request:
### Description of Change ###

As proposed in #2060, introduced a MaxLines property on a `Label`. Implemented for:
- iOS
- Mac OS
- Android
- UWP

### Issues Resolved ###

- fixes #2060
- fixes #1706

### API Changes ###

Added:
 - int Label.MaxLines { get; set; } //Bindable Property

### Platforms Affected ###

<!-- Please list all platforms affected by these changes -->

- Core/XAML (all platforms)
- iOS
- Mac OS
- Android
- UWP

### Behavioral/Visual Changes ###

Since this is a new property with a default value as the control has now, the user shouldn't see any changes. When setting the property, the `Label` will limit the number of lines to the stated value.

### PR Checklist ###

- [ ] Has automated tests <!-- (if tests are omitted or manual, state reason in description) -->
- [X] Rebased on top of the target branch at time of PR
- [X] Changes adhere to coding standard

jfversluis commented:
Not sure why the maxlines in the iOS/Mac OS renderer were set when setting the truncation. From what I tested this works together well the way I implemented it now. But a second-opinion is appreciated.

jfversluis commented:
Also, if anyone has any idea on how to implement on GTK and Tizen, I'm open to that.

jfversluis commented:
Uh-oh. What did I break?

StephaneDelcroix commented on Xamarin.Forms.Core/Label.cs:
```diff
@@ -78,6 +78,8 @@ public class Label : View, IFontElement, ITextElement, ITextAlignmentElement, IL
 
 		public static readonly BindableProperty LineHeightProperty = LineHeightElement.LineHeightProperty;
 
+		public static readonly BindableProperty MaxLinesProperty = BindableProperty.Create(nameof(MaxLines), typeof(int), typeof(Label), default(int));
```
use `-1` has default

jfversluis replied:
Done :)

StephaneDelcroix commented on Xamarin.Forms.Platform.Android/Renderers/LabelRenderer.cs:
```diff
@@ -197,6 +202,15 @@ void UpdateLineHeight()
+			{
+				Control.SetSingleLine(Element.MaxLines == 1);
+				Control.SetMaxLines(Element.MaxLines);
+			}
```
a `else` is missing. it should handled the case where a `MaxLines` value is set, then `ClearValue()`'d, actually reverting to the control default.

StephaneDelcroix commented on Xamarin.Forms.Platform.UAP/LabelRenderer.cs:
```diff
@@ -312,5 +315,13 @@ void UpdateLineHeight(TextBlock textBlock)
+			if (Element.MaxLines >= 0)
+			{
+				textBlock.MaxLines = Element.MaxLines;
+			}
```
same

StephaneDelcroix commented on Xamarin.Forms.Platform.iOS/Renderers/LabelRenderer.cs:
```diff
@@ -334,5 +325,21 @@ void UpdateLayout()
+
+				Layout();
+#endif
+			}
```
same

jfversluis commented:
Rookie mistake... Added it now!

Also added `>= 0` check on iOS and Mac instead of `> 0` because setting the value to 0 sets it to auto-size. Might be nice to include that for our users.

StephaneDelcroix commented on Xamarin.Forms.Core/Label.cs:
```diff
@@ -78,6 +78,8 @@ public class Label : View, IFontElement, ITextElement, ITextAlignmentElement, IL
 
 		public static readonly BindableProperty LineHeightProperty = LineHeightElement.LineHeightProperty;
 
+		public static readonly BindableProperty MaxLinesProperty = BindableProperty.Create(nameof(MaxLines), typeof(int), typeof(Label), -1);
```
we probably should InvalidateMeasure in case the property changes.

also, there's some code in `OnTextChanged()` that rely on some heuristics to find out if the label is single line or not. those heuristics aren't probably valid anymore.

StephaneDelcroix commented on Xamarin.Forms.Platform.Android/Renderers/LabelRenderer.cs:
```diff
@@ -197,6 +202,20 @@ void UpdateLineHeight()
+			else
+			{
+				Control.SetSingleLine(true);
+				Control.SetMaxLines(1);
```
I'm might be wrong here, but I'm not sure the default is single line labels.

jfversluis replied:
Correct, will change

StephaneDelcroix commented on Xamarin.Forms.Platform.UAP/LabelRenderer.cs:
```diff
@@ -312,5 +315,17 @@ void UpdateLineHeight(TextBlock textBlock)
+			}
+			else
+			{
+				textBlock.MaxLines = 0;
```
0 means 'no maximum', right ?

jfversluis replied:
0 == auto and is the default value: https://docs.microsoft.com/en-us/uwp/api/windows.ui.xaml.controls.textblock.maxlines#Windows_UI_Xaml_Controls_TextBlock_MaxLines

StephaneDelcroix commented on Xamarin.Forms.Platform.iOS/Renderers/LabelRenderer.cs:
```diff
@@ -334,5 +325,33 @@ void UpdateLayout()
+#if __MOBILE__
+				Control.Lines = 1;
+
+				LayoutSubviews();
```
it probably depends on the LineBreakMode

jfversluis replied:
Default value == 1, as per: https://developer.apple.com/documentation/uikit/uilabel/1620539-numberoflines

jfversluis replied:
Default value == 1, as per: https://developer.apple.com/documentation/uikit/uilabel/1620539-numberoflines.

But agreed that using in conjunction with line break mode makes sense. Then 1 for truncation and 0 (which acts as auto) for wrapping?

StephaneDelcroix replied:
> Then 1 for truncation and 0 (which acts as auto) for wrapping?

that's the way it was, iirc

jfversluis replied:
You're right! 🙂 pfew, have to get back in there. Was a while ago when I drafted this

jfversluis commented:
Updates again!

jfversluis commented:
Ah, I should've mentioned. I didn't think so, because that's kind of the same as this logic on iOS:
https://github.com/xamarin/Xamarin.Forms/pull/3318/commits/1ae5df8523d2d398b1bd38fb91d50927e2767df4#diff-381c6c37e826f38028e8841495e4f78eR364

`CharacterWrap` and `WordWrap` mean multi-line, else single line. Still seems to apply?

StephaneDelcroix commented:
Thanks for this PR, @jfversluis

jfversluis commented:
No problem at all! Happy to contribute 👍 
Thank _you_ for all the good feedback and guidance!
